articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/**:
+    ros__qos_profiles:
+        profiles:
+            reliable_depth_100:
+                reliability: reliable
+                history_depth: 100
+                history: keep_last
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                base: reliable_depth_100
+        publisher:
+            topic_name: bsd
+            qos:
+                base: reliable_depth_100
+                lifespan: 10s
+```
+
+This mechanism will help ensuring matching QoS profile, where that's needed in the system.
+
+Having some implicitly defined named profiles can become handy to easily write profiles.
+The following could be implicitly defined:
+
+- ros_default: Equivalent to `rmw_qos_profile_default`.
+- ros_sensor_data: Equivalent to `rmw_qos_profile_sensor_data`.
+- ros_service_default: Equivalent to `rmw_qos_profile_services_default` (as of ROS 2 Foxy, the same as `ros_default`).
+- ros_system_default: Equivalent to `rmw_qos_profile_system_default`.
+
+See [qos_profiles.h](https://github.com/ros2/rmw/blob/master/rmw/include/rmw/qos_profiles.h).
+
+## Implicit base profile
+
+It might happen that a QoS profile is not fully specified and no base profile was chosen, in that case there will be an implicit QoS profile.
+If there was a profile defined in the `<default>` section for the same topic or the same service, that one will be used:
+
+```xml
+<qos_profiles>
+    <default>
+        <publisher topic_name="/my_ns/nested_ns/asd">
+            <qos>
+                <reliability>reliable</reliability>
+                <history_depth>100</history_depth>
+            </qos>
+        </publisher>
+    </default>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <history_depth>1000</history_depth>  <!--Overrides the history depth of 100 defined above.-->
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/**:
+    ros__qos_profiles:
+        publisher:
+            topic_name: /my_ns/nested_ns/asd
+            qos:
+                reliability: reliable
+                history_depth: 100
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                history_depth: 1000
+```
+
+If that doesn't happen, `rmw_qos_profile_default` is used as a base.
+
+### Rationale
+
+There are other options to handle a non fully specified qos profile:
+
+- Reject those qos profiles, and force users to specify all policies.
+- Use the original qos profile specified in code as a base profile.
+
+The first option would be extremely verbose, and it doesn't make much sense to explicitly set some QoS settings like `lifespan` when you don't care about them.
+The last option is possible, but it would be hard to tell what was the original profile, and code would have been to be carefully checked.
```

Comment:
Yeah, I originally was going to propose this alternative, but then I changed my mind.
I think there are two cases:
1. Profiles in the `<default>` or `<profiles>` section.
2. Profiles defined inline in a `<node>` section.

In case 1., using the QoS profile specified by the author as a base doesn't make sense.
Imagine the following case:

```xml
<qos_profiles>
    <default>
        <publisher topic_name="/my_ns/nested_ns/asd">
            <qos>
              <history_depth>1000</history_depth>
            </qos>
        </publisher>
    </default>
</qos_profiles>
```

Node A has reliable QoS policy set for its /my_ns/nested_ns/asd publisher in code.
Node B has best effor QoS policy set for its /my_ns/nested_ns/asd publisher in code.

Is the profile in code is taken as the default, it will create an unexpected difference between the two profiles.
So maybe for the profiles in the `<default>` section, my suggestion makes sense.

The problem is similar with "named profiles".

Maybe `base=<profile_name>`  should be mandatory in this case, that's not a bad idea.

For case 2., e.g:

```xml
<qos_profiles>
    <node name=... namespace=...>
        <publisher topic_name="/my_ns/nested_ns/asd">
            <qos>
              <history_depth>1000</history_depth>
            </qos>
        </publisher>
    </node>
</qos_profiles>
```

your suggestion makes sense to me.

What do you think?

Reply:
Yes, that makes sense to me. Basically, we should update the line above:

```diff
- If that doesn't happen, `rmw_qos_profile_default` is used as a base.
+If no base profile is specified and no default profile exists for an entity, then the QoS settings set by the node author are used as a base.
```

Reply:
Solved in https://github.com/ros2/design/pull/296/commits/94c1e103ba494263440b7a7d451762b88b494942.